### DISCUSSION TITLE
multi language app, in the route

### DISCUSSION THREAD

**Discussion by leognmotta** [TanStack/router#1841] [Q&A]
I looking forward how to achieve internationalization on the router like:

```
/renders-app
/pt/renders-app
```

basically optionally adding a prefix on the path, how could I get this done using file based router?

**Comment by TacticsJan** [comment]
Also am interested in knowing how this can be achieved with Tanstack Router

**Comment by stevenlyd** [comment]
I guess you can use route masking?
https://tanstack.com/router/latest/docs/framework/react/guide/route-masking

**Reply by samuelstroschein** [reply]
Tried to get [Paraglide JS](https://inlang.com/m/gerre34r/library-inlang-paraglideJs) to work with masking but masking doesn't behave as expected. 

The from path is not valid (does the route have to exist?). A 404 pops up. Anyone investigating this further should create "virtual routes" for the localized paths via [localizeHref()](https://inlang.com/m/gerre34r/library-inlang-paraglideJs/runtime#localizehref) which are then mapped back to the original path for tanstack router to render via [deLocalizeHref()](https://inlang.com/m/gerre34r/library-inlang-paraglideJs/runtime#delocalizehref).

```ts
const localizedRouteMask = createRouteMask({
	routeTree: routeTree,
	from: "/de/",
	to: "/",
});

console.log(routeTree);

// Set up a Router instance
const router = createRouter({
	routeTree,
	defaultPreload: "intent",
	routeMasks: [localizedRouteMask],
});
```

**Reply by EliasGit117** [reply]
@samuelstroschein have you managed to fix that?

**Reply by samuelstroschein** [reply]
@EliasGit117 no

**Reply by EliasGit117** [reply]
@samuelstroschein, I have an working solution, do you want me to share?

**Reply by aabuhijleh** [reply]
@EliasGit117 I'm interested as well. Please share your solution.

**Reply by EliasGit117** [reply]
```
plugins: [
  TanStackRouterVite({ target: 'react', autoCodeSplitting: true }),
  svgr(),
  react(),
  tailwindcss(),
  paraglideVitePlugin({
    project: './project.inlang',
    outdir: './src/paraglide',
    cookieName: 'lang',
    localStorageKey: 'lang',
    strategy: ['url', 'cookie', 'preferredLanguage', 'baseLocale'],
  }),
],

  
// Set up a Router instance
const router = createRouter({
  routeTree,
  context: {
    queryClient,
  },
  defaultPreload: 'intent',
  // Since we're using React Query, we don't want loader calls to ever be stale
  // This will ensure that the loader is always called when the route is preloaded or visited
  scrollRestoration: true,
  defaultPendingMinMs: 0,
  basepath: getRouterBasepath(window.location.pathname),
})



export function extractLocale(url: string): Locale | undefined {
  const urlObj = new URL(url, 'http://dummy.com');
  const pathSegments = urlObj.pathname.split('/').filter(Boolean);
  if (pathSegments.length > 0) {
    const potentialLocale = pathSegments[0];
    if (isLocale(potentialLocale)) {
      return potentialLocale;
    }
  }
}


export function getRouterBasepath(pathname?: string): string | undefined {
  let locale = extractLocale(pathname ?? '/');

  return (strategy.includes('url') && locale) ? `/${locale}` : undefined;
}


}

```

**Reply by samuelstroschein** [reply]
@EliasGit117 why don't you use the `detectLocaleFromUrl` function of paraglide js? I assume because you don't have access to the API in the vite config (it's not compiled yet)?

**Reply by shlroland** [reply]
> ```
> plugins: [
>   TanStackRouterVite({ target: 'react', autoCodeSplitting: true }),
>   svgr(),
>   react(),
>   tailwindcss(),
>   paraglideVitePlugin({
>     project: './project.inlang',
>     outdir: './src/paraglide',
>     cookieName: 'lang',
>     localStorageKey: 'lang',
>     strategy: ['url', 'cookie', 'preferredLanguage', 'baseLocale'],
>   }),
> ],
> 
>   
> // Set up a Router instance
> const router = createRouter({
>   routeTree,
>   context: {
>     queryClient,
>   },
>   defaultPreload: 'intent',
>   // Since we're using React Query, we don't want loader calls to ever be stale
>   // This will ensure that the loader is always called when the route is preloaded or visited
>   scrollRestoration: true,
>   defaultPendingMinMs: 0,
>   basepath: getRouterBasepath(window.location.pathname),
> })
> 
> 
> 
> export function extractLocale(url: string): Locale | undefined {
>   const urlObj = new URL(url, 'http://dummy.com');
>   const pathSegments = urlObj.pathname.split('/').filter(Boolean);
>   if (pathSegments.length > 0) {
>     const potentialLocale = pathSegments[0];
>     if (isLocale(potentialLocale)) {
>       return potentialLocale;
>     }
>   }
> }
> 
> 
> export function getRouterBasepath(pathname?: string): string | undefined {
>   let locale = extractLocale(pathname ?? '/');
> 
>   return (strategy.includes('url') && locale) ? `/${locale}` : undefined;
> }
> 
> 
> }
> ```

I want to know how you switch languages.

**Reply by EliasGit117** [reply]
@shlroland  setLocale(locale)

**Reply by shlroland** [reply]
> @shlroland setLocale(locale)

My meaning is, after switching languages, will the basePath change, and how should the routing respond to this change?I didn't use paraglidejs, it was another intl library. I'm surprised that tanstack router doesn't provide best practices in this regard

**Reply by EliasGit117** [reply]
@shlroland it will change if you have strategy “ur” in config of paraglide, or if you overwrite setLocale (check paraglide docs)

**Reply by rxliuli** [reply]
I am using i18next and cannot find any available community templates. The main requirement is to support language prefixes like /en /ja and generate SSG static websites.

**Comment by EliasGit117** [comment]
@samuelstroschein because it returns base locale if url is empty, I'm trying to implement that in case I don't have locale in url it should use value from cookies

```
overwriteGetLocale(() => {
  const fromUrl =  extractLocale(window.location.pathname);

  if (!!fromUrl)
    return fromUrl;

  const fromCookies = extractLocaleFromCookie();
  if (isLocale(fromCookies))
    return fromCookies;

  return baseLocale;
})
```

```
strategy: ['cookie', 'url', 'preferredLanguage', 'baseLocale'],
```

**Comment by 58bits** [comment]
We'll be using optional parameters for this. There's a very good description of how to do this in the docs....

https://tanstack.com/router/v1/docs/framework/react/guide/path-params#internationalization-i18n-with-optional-path-parameters

**Reply by murbanowicz** [reply]
Any idea how to handle translated paths? Optional params can address lot, but having multiple languages pointing eg /pricing.tsx is so far impossible for me. In theory paraglide exposes some middleware for this but it doesn't work

**Comment by samuelstroschein** [comment]
@juliomuhlbauer created examples for TanStack Router and Start which supports i18n routing via Paraglide JS 

- [TanStack Router example](https://github.com/TanStack/router/tree/main/examples/react/i18n-paraglide)
- [TanStack Start example](https://github.com/TanStack/router/tree/main/examples/react/start-i18n-paraglide)
